src/langchain_google_spanner/graph_qa.py

```diff
@@ -0,0 +1,375 @@
+            except Exception as e:
+                err_msg = str(e)
+                self.log_invalid_query(_run_manager, gql_query, err_msg)
+                intermediate_steps.pop()
+                intermediate_steps.append({"query_failed_" + str(retries): gql_query})
+                fix_chain_result = self.gql_fix_chain.invoke(
+                    {
+                        "question": question,
+                        "err_msg": err_msg,
+                        "generated_gql": gql_query,
+                        "schema": self.graph.get_schema,
+                    }
+                )
+                gql_query = extract_gql(fix_chain_result)
+            finally:
+                retries += 1
+
+        raise ValueError("The generated gql query is invalid")
+
+    def log_invalid_query(
+        self,
+        _run_manager: CallbackManagerForChainRun,
+        generated_gql: str,
+        err_msg: str,
+    ) -> None:
+        _run_manager.on_text("Invalid generated gql:", end="\n", verbose=self.verbose)
+        _run_manager.on_text(generated_gql, color="red", end="\n", verbose=self.verbose)
+        _run_manager.on_text(
+            "Query error: ", color="red", end="\n", verbose=self.verbose
+        )
+        _run_manager.on_text(err_msg, color="red", end="\n", verbose=self.verbose)
+
+    def _call(
+        self,
+        inputs: Dict[str, Any],
+        run_manager: Optional[CallbackManagerForChainRun] = None,
+    ) -> Dict[str, str]:
+
+        intermediate_steps: List = []
+
+        """Generate gql statement, uses it to look up in db and answer question."""
+
+        _run_manager = run_manager or CallbackManagerForChainRun.get_noop_manager()
+        question = inputs[self.input_key]
+        gen_response = self.gql_generation_chain.invoke(
+            {"question": question, "schema": self.graph.get_schema},
+        )
+        generated_gql = extract_gql(gen_response)
+
+        if self.verify_gql:
+            verify_response = self.gql_verify_chain.invoke(
+                {
+                    "question": question,
+                    "generated_gql": generated_gql,
+                    "graph_schema": self.graph.get_schema,
+                }
+            )
+            verified_gql = fix_gql_syntax(verify_response["verified_gql"])
+            intermediate_steps.append({"verified_gql": verified_gql})
+        else:
+            verified_gql = generated_gql
+
+        final_gql = ""
+        if verified_gql:
+            (final_gql, context) = self.execute_with_retry(
+                _run_manager, intermediate_steps, question, verified_gql
+            )
+            _run_manager.on_text("Full Context:", end="\n", verbose=self.verbose)
+            _run_manager.on_text(
+                str(context), color="green", end="\n", verbose=self.verbose
+            )
+            intermediate_steps.append({"context": context})
+        else:
```

Comment:
if `final_gql` is empty string, should we return error?

Reply:
done